File: day-09/day-09.py
with open('data.txt', 'r') as r:
    nums = r.read()

data = []
dots = []
curID = 0
for i in range(len(nums)):
    num = int(nums[i])
    if i % 2 == 0:
        data.extend([curID] * num)
        curID += 1
    else:
        data.extend(['.'] * num)
        dots.extend(range(len(data)-num, len(data)))

data2 = data.copy()

for idx in dots:
    if idx > len(data):
        break
    val = data.pop()
    data[idx] = val
    while data[-1] == '.':
        data.pop()

part1 = 0
for i, e in enumerate(data):
    part1 += int(e) * i

print("Part 1:", part1)

# Part 2

dot_groups = []
for i in range(len(data2)):
    if data2[i] == '.':
        if i == 0 or data2[i - 1] != '.':
            dot_groups.append([i])
        else:
            dot_groups[-1].append(i)

file_data = []
i = 0
while i < len(data2):
    if data2[i] != '.':
        file_id = data2[i]
        start = i
        while i < len(data2) and data2[i] == file_id:
            i += 1
        size = i - start
        file_data.append((file_id, size, start))
    else:
        i += 1

file_data = file_data[::-1]
for file_id, size, start_index in file_data:
    for j, dots in enumerate(dot_groups):
        if len(dots) >= size and dots[0] < start_index:  
            for k in range(size):
                data2[dots[k]] = file_id
                data2[start_index + k] = '.'
            dot_groups[j] = dots[size:]
            if len(dot_groups[j]) == 0:
                dot_groups.pop(j)
            break

part2 = 0
for i, e in enumerate(data2):
    if e != '.':
        part2 += int(e) * i

print("Part 2:", part2)

File: day-09/day-09.rs
use std::fs::File;
use std::io::{self, Read};
use std::vec;

fn main() -> io::Result<()> {
    // Read the file
    let mut file = File::open("data.txt")?;
    let mut nums = String::new();
    file.read_to_string(&mut nums)?;

    // Part 1 and Part 2 processing
    let (part1, part2) = process_file(&nums);

    println!("Part 1: {}", part1);
    println!("Part 2: {}", part2);

    Ok(())
}

fn process_file(nums: &str) -> (usize, usize) {
    let mut data = Vec::new();
    let mut dots = Vec::new();
    let mut cur_id = 0;

    // First pass: populate data and dots
    for (i, ch) in nums.chars().enumerate() {
        let num = ch.to_digit(10).unwrap() as usize;
        if i % 2 == 0 {
            // File IDs
            data.extend(std::iter::repeat(cur_id).take(num));
            cur_id += 1;
        } else {
            // Dot placeholders
            let start = data.len();
            data.extend(std::iter::repeat('.').take(num));
            dots.extend(start..data.len());
        }
    }

    // Part 1 processing
    let mut data_part1 = data.clone();
    for idx in dots.iter() {
        if *idx >= data_part1.len() {
            break;
        }
        let val = data_part1.pop().unwrap();
        data_part1[*idx] = val;
    }

    // Trim trailing dots
    while let Some('.') = data_part1.last() {
        data_part1.pop();
    }

    let part1 = data_part1.iter().enumerate()
        .map(|(i, &e)| if e == '.' { 0 } else { i * e })
        .sum();

    // Part 2 processing
    let mut data_part2 = data.clone();
    let mut dot_groups = Vec::new();

    // Identify dot groups
    for (i, &item) in data_part2.iter().enumerate() {
        if item == '.' {
            if i == 0 || data_part2[i - 1] != '.' {
                dot_groups.push(vec![i]);
            } else {
                dot_groups.last_mut().unwrap().push(i);
            }
        }
    }

    // Identify file data
    let mut file_data = Vec::new();
    let mut i = 0;
    while i < data_part2.len() {
        if data_part2[i] != '.' {
            let file_id = data_part2[i];
            let start = i;
            while i < data_part2.len() && data_part2[i] == file_id {
                i += 1;
            }
            let size = i - start;
            file_data.push((file_id, size, start));
        } else {
            i += 1;
        }
    }
    file_data.reverse();

    // Place files
    for (file_id, size, start_index) in file_data {
        for (j, dots) in dot_groups.iter_mut().enumerate() {
            if dots.len() >= size && dots[0] < start_index {
                for k in 0..size {
                    data_part2[dots[k]] = file_id;
                    data_part2[start_index + k] = '.';
                }
                *dots = dots[size..].to_vec();
                if dots.is_empty() {
                    dot_groups.remove(j);
                }
                break;
            }
        }
    }

    // Calculate part 2
    let part2 = data_part2.iter().enumerate()
        .map(|(i, &e)| if e == '.' { 0 } else { i * e })
        .sum();

    (part1, part2)
}